sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/AdaptiveSparkPlanExec.scala

```diff
@@ -0,0 +1,383 @@
+  }
+
+  private def reuseQueryStage(s: QueryStageExec, output: Seq[Attribute]): QueryStageExec = {
+    val queryStage = ReusedQueryStageExec(currentStageId, s, output)
+    currentStageId += 1
+    queryStage
+  }
+
+  /**
+   * Returns the updated logical plan after new query stages have been created and the physical
+   * plan has been updated with the newly created stages.
+   * 1. If the new query stage can be mapped to an integral logical sub-tree, replace the
+   *    corresponding logical sub-tree with a leaf node [[LogicalQueryStage]] referencing the new
+   *    query stage. For example:
+   *        Join                   SMJ                      SMJ
+   *      /     \                /    \                   /    \
+   *    r1      r2    =>    Xchg1     Xchg2    =>    Stage1     Stage2
+   *                          |        |
+   *                          r1       r2
+   *    The updated plan node will be:
+   *                               Join
+   *                             /     \
+   *    LogicalQueryStage1(Stage1)     LogicalQueryStage2(Stage2)
+   *
+   * 2. Otherwise (which means the new query stage can only be mapped to part of a logical
+   *    sub-tree), replace the corresponding logical sub-tree with a leaf node
+   *    [[LogicalQueryStage]] referencing to the top physical node into which this logical node is
+   *    transformed during physical planning. For example:
+   *     Agg           HashAgg          HashAgg
+   *      |               |                |
+   *    child    =>     Xchg      =>     Stage1
+   *                      |
+   *                   HashAgg
+   *                      |
+   *                    child
+   *    The updated plan node will be:
+   *    LogicalQueryStage(HashAgg - Stage1)
+   */
+  private def updateLogicalPlan(
+      logicalPlan: LogicalPlan,
+      newStages: Seq[(Exchange, QueryStageExec)]): LogicalPlan = {
+    var currentLogicalPlan = logicalPlan
+    newStages.foreach { case (exhange: Exchange, stage: QueryStageExec) =>
+      // Get the corresponding logical node for `exchange`. If `exchange` has been transformed from
+      // a `Repartition`, it should have `logicalLink` available by itself; otherwise traverse down
+      // to find the first node that is not generated by `EnsureRequirements`.
+      val logicalNodeOpt = exhange.logicalLink.orElse(exhange.collectFirst {
+        case p if p.logicalLink.isDefined => p.logicalLink.get
+      })
+      assert(logicalNodeOpt.isDefined)
+      val logicalNode = logicalNodeOpt.get
+      val physicalNode = currentPhysicalPlan.collectFirst {
+        case p if p.eq(stage) || p.logicalLink.exists(logicalNode.eq) => p
+      }
+      assert(physicalNode.isDefined)
+      // Replace the corresponding logical node with LogicalQueryStage
+      val newLogicalNode = LogicalQueryStage(logicalNode, physicalNode.get)
+      // Force update logical link
+      physicalNode.get.setTagValue(SparkPlan.LOGICAL_PLAN_TAG, newLogicalNode)
```

Comment:
do we need to do this? I think we will update the link automatically when the `LogicalQueryStageStrategy` is executed by the planner.

Reply:
The problem was the original `SparkPlan.setLogicalLink` did not "force set" it. I've changed it now.